keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlResourceRepository.kt

```diff
@@ -141,18 +143,53 @@ class SqlResourceRepository(
     }
   }
 
+  override fun nextResourcesDueForCheck(minTimeSinceLastCheck: Duration, limit: Int): Collection<ResourceHeader> {
+    val now = clock.instant()
+    val cutoff = now.minus(minTimeSinceLastCheck).let(Timestamp::from)
+    return jooq.inTransaction {
+      select(UID, API_VERSION, KIND, NAME)
+        .from(RESOURCE)
+        .where(LAST_CHECKED.isNull.or(LAST_CHECKED.lessOrEqual(cutoff)))
+        .orderBy(LAST_CHECKED)
+        .limit(limit)
+        .forUpdate()
```

Comment:
I suspect this would result in deadlocks w/mysql's default isolation level but I think it will be ok in `READ_COMMITTED` which kork-sql should already be setting on connections by default (it's already the default postgres isolation level.) I'd still expect lock wait contention between concurrently scheduled keel instances which may constrain how large the `limit` variable here can be set to. I can imagine a world where we have a resource table per type, with checks scheduled independently behind a short lived distributed lock per type to get better throughput, though it would be nice if small limits avoid the need for this without throughput being an issue.

To help ascertain how viable this is, we can do some benchmarking via an expanded version of `SqlResourceRepositoryTests.parallelCheckingTests()` with millions of rows and iterations testing increasing `limit` values. It would also be good to include some concurrent inserts of new resources to validate that insert performance isn't impacted. This is going to need some db configuration to actually get useful lock wait timing data. It could also be run against an Aurora instance that has Perf Insights enabled, in which case we'd be able to graph lock wait times in great detail.

Reply:
I was thinking of using a limit of 1 to start with!